src/libsystemd-network/sd-dhcp-lease.c

```diff
@@ -559,6 +572,218 @@ static int lease_parse_sip_server(const uint8_t *option, size_t len, struct in_a
+                        if (!make_cstring((const char*) &option[offset], plen,
+                                        MAKE_CSTRING_REFUSE_TRAILING_NUL, &dohpath))
+                                return -EINVAL;
+                        break;
+
+                default:
+                        break;
+                }
+                offset += plen;
+        }
+        if (offset != len)
+                return -EINVAL;
+
+        /* DNR cannot be used without alpn */
+        if (!alpn)
+                return -EINVAL;
+
+        /* DoH cannot be used without dohpath */
+        if (!dohpath)
+                SET_FLAG(transports, SD_DNS_ALPN_HTTP_2_TLS|SD_DNS_ALPN_HTTP_3, false);
+
+        /* No useful transports */
+        if (!transports)
+                return 0;
+
+        resolver->transports = transports;
+        resolver->port = port;
+        free_and_replace(resolver->dohpath, dohpath);
+        return transports;
+}
+
+static int lease_parse_designated_resolvers(const uint8_t *option, size_t len, ResolverData **ret_resolvers,
+                size_t *ret_n_resolvers) {
+        int r;
+        _cleanup_(dhcp_resolver_data_free_allp) ResolverData *resolvers = NULL;
+        size_t n_resolvers = 0;
+
+        assert(option);
+        assert(ret_resolvers);
+        assert(ret_n_resolvers);
+
+        size_t offset = 0;
+        while (offset < len) {
+                _cleanup_(dhcp_resolver_data_free_allp) ResolverData *resolver = NULL;
+                if (len < 2)
+                        return -EINVAL;
+
+                /* Instance Data length */
+                size_t ilen = unaligned_read_be16(&option[offset]);
+                if (2 + ilen > len)
+                        return -EINVAL;
+                offset += 2;
+
+                resolver = new0(ResolverData, 1);
+                if (!resolver)
+                        return -ENOMEM;
+
+                resolver->priority = unaligned_read_be16(&option[offset]);
+                offset += 2;
+
+                /* Authenticated Domain Name */
+                ilen = option[offset++];
+                if (offset + ilen > len)
+                        return -EINVAL;
+
+                r = lease_parse_dns_name(&option[offset], ilen, &resolver->auth_name);
+                if (r < 0)
+                        return r;
+                offset += ilen;
+
+                /* IPv4 addrs */
+                ilen = option[offset++];
+                r = lease_parse_in_addrs(&option[offset], ilen, &resolver->addrs, &resolver->n_addrs);
+                if (r < 0)
+                        return r;
+                offset += ilen;
+
+                /* service params */
+                r = lease_parse_dnr_params(&option[offset], len-offset, resolver);
+                if (r < 0)
+                        return r;
+                else if (r == 0) {
+                        *ret_resolvers = dhcp_resolver_data_free_all(*ret_resolvers);
```

Comment:
if you both read from and write to a parameter it is not an output parameter, and hence should not be named `ret_xyz`